homeassistant/components/tplink/camera.py

```diff
@@ -0,0 +1,156 @@
+"""Support for TPLink camera entities."""
+
+import asyncio
+from dataclasses import dataclass
+import logging
+
+from aiohttp import web
+from haffmpeg.camera import CameraMjpeg
+from kasa import Device, Module
+from kasa.smartcam.modules import Camera as CameraModule
+
+from homeassistant.components import ffmpeg
+from homeassistant.components.camera import (
+    Camera,
+    CameraEntityDescription,
+    CameraEntityFeature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.aiohttp_client import async_aiohttp_proxy_stream
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import TPLinkConfigEntry, legacy_device_id
+from .coordinator import TPLinkDataUpdateCoordinator
+from .entity import CoordinatedTPLinkEntity, TPLinkModuleEntityDescription
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class TPLinkCameraEntityDescription(
+    CameraEntityDescription, TPLinkModuleEntityDescription
+):
+    """Base class for camera entity description."""
+
+
+CAMERA_DESCRIPTIONS: tuple[TPLinkCameraEntityDescription, ...] = (
+    TPLinkCameraEntityDescription(
+        key="live_view",
+        translation_key="live_view",
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: TPLinkConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up camera entities."""
+    data = config_entry.runtime_data
+    parent_coordinator = data.parent_coordinator
+    device = parent_coordinator.device
+    ffmpeg_manager = ffmpeg.get_ffmpeg_manager(hass)
+
+    async_add_entities(
+        TPLinkCameraEntity(
+            device,
+            parent_coordinator,
+            description,
+            camera_module=camera_module,
+            parent=None,
+            ffmpeg_manager=ffmpeg_manager,
+        )
+        for description in CAMERA_DESCRIPTIONS
+        if (camera_module := device.modules.get(Module.Camera))
+    )
+
+
+class TPLinkCameraEntity(CoordinatedTPLinkEntity, Camera):
+    """Representation of a TPLink camera."""
+
+    _attr_supported_features = CameraEntityFeature.STREAM | CameraEntityFeature.ON_OFF
+
+    entity_description: TPLinkCameraEntityDescription
+
+    def __init__(
+        self,
+        device: Device,
+        coordinator: TPLinkDataUpdateCoordinator,
+        description: TPLinkCameraEntityDescription,
+        *,
+        camera_module: CameraModule,
+        parent: Device | None = None,
+        ffmpeg_manager: ffmpeg.FFmpegManager,
+    ) -> None:
+        """Initialize a TPlink camera."""
+        self.entity_description = description
+        self._camera_module = camera_module
+        self._video_url: str | None = None
+        self._image: bytes | None = None
+        super().__init__(device, coordinator, parent=parent)
+        Camera.__init__(self)
+        self._ffmpeg_manager = ffmpeg_manager
+        self._image_lock = asyncio.Lock()
```

Comment:
```suggestion
        self._ffmpeg_manager = ffmpeg_manager
        self._image_lock = asyncio.Lock()
        super().__init__(device, coordinator, parent=parent)
        Camera.__init__(self)
```
Or probably more preferably, the super class initializations should be done in the beginning of the ctor?

Reply:
See https://github.com/home-assistant/core/pull/129180#discussion_r1818137159. Moving the super to the top requires a bigger refactoring because of when we call `_async_call_update_attrs`.